src/v/iceberg/filesystem_catalog.cc

```diff
@@ -0,0 +1,240 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "iceberg/filesystem_catalog.h"
+
+#include "iceberg/manifest_entry.h"
+#include "iceberg/table_metadata.h"
+#include "iceberg/table_update_applier.h"
+#include "model/timestamp.h"
+
+namespace iceberg {
+
+namespace {
+catalog::errc to_catalog_errc(metadata_io::errc e) {
+    switch (e) {
+    case metadata_io::errc::shutting_down:
+        return catalog::errc::shutting_down;
+    case metadata_io::errc::timedout:
+        return catalog::errc::timedout;
+    case metadata_io::errc::failed:
+        return catalog::errc::io_error;
+    }
+}
+constexpr std::string_view vhint_filename = "version-hint.text";
+ss::sstring tmeta_filename(int version) {
+    return fmt::format("v{}.metadata.json", version);
+}
+} // namespace
+
+ss::future<checked<table_metadata, catalog::errc>>
+filesystem_catalog::create_table(
+  const table_identifier& table_ident,
+  const schema& s,
+  const partition_spec& spec) {
+    // TODO: at some level (maybe the caller), we need to manually re-assign
+    // determininstic, unique field IDs, and re-assign the partition ids. The
+    // Java library and the REST catalog will reassign these on their end, and
+    // we need to follow suit, lest we end up with a mismatch.
+
+    auto new_schema = s.copy();
+    new_schema.schema_id = schema::id_t{0};
+    auto highest_column_id = new_schema.highest_field_id().value_or(
+      nested_field::id_t{0});
+    auto new_spec = spec.copy();
+    new_spec.spec_id = partition_spec::id_t{0};
+    // NOTE: by convention, the first partition field should be id 1000.
+    auto highest_pid = partition_field::id_t{999};
+    for (const auto& p : new_spec.fields) {
+        highest_pid = std::max(highest_pid, p.field_id);
+    }
+
+    // Construct a metadata with all initial values.
+    chunked_vector<sort_order> sort_orders;
+    sort_orders.emplace_back(
+      sort_order{sort_order::id_t{0}, chunked_vector<sort_field>{}});
+    chunked_vector<schema> schemas;
+    schemas.emplace_back(std::move(new_schema));
+    chunked_vector<partition_spec> specs;
+    specs.emplace_back(std::move(new_spec));
```

Comment:
Is it possible to create these three vectors in place?

Reply:
Done

Reply:
Ah actually not, since initializer lists can only be copied and the underlying types here all have fragmented vectors, which aren't default copyable